fem/pmgbilinearform.hpp

```diff
@@ -0,0 +1,44 @@
+// Copyright (c) 2010, Lawrence Livermore National Security, LLC. Produced at
+// the Lawrence Livermore National Laboratory. LLNL-CODE-443211. All Rights
+// reserved. See file COPYRIGHT for details.
+//
+// This file is part of the MFEM library. For more information and source code
+// availability see http://mfem.org.
+//
+// MFEM is free software; you can redistribute it and/or modify it under the
+// terms of the GNU Lesser General Public License (as published by the Free
+// Software Foundation) version 2.1 dated February 1999.
+
+#ifndef MFEM_PMGBILINEARFORM
+#define MFEM_PMGBILINEARFORM
+
+#include "../linalg/multigrid.hpp"
+#include "spacehierarchy.hpp"
+
+#include "mgbilinearform.hpp"
+#include "pbilinearform.hpp"
+
+namespace mfem
+{
+
+class ParMultigridBilinearForm : public MultigridBilinearForm
```

Comment:
The same comments from `MultigridBilinearForm` also apply to this class.

It would also be great to avoid the code duplication associated with the `Par` objects. I seem to remember that you originally had a templated version, and changed to this non-templated version to make the code more in line with the MFEM style.

I don't have a great solution for this other than templates (perhaps we should consider using templates in this case), but I do think that the code duplication is an anti-pattern, which will make code maintenance more difficult and increase the likelihood of bugs.

Reply:
You are right. These and the `SpaceHierarchy` classes were originally implemented using templates. Does anyone else have a preference of what to use in this case?

Reply:
We can have one object for serial and parallel. If using just virtual methods from `FiniteElementSpace` or `BilinearForm` there will be no need to write different code in serial and parallel. If we need to distinguish the serial/parallel case we can use `dynamic_cast`, e.g. from `FiniteElementSpace*` to `ParFiniteElementSpace*`. This approach is used e.g. in class `StaticCondensation`.

Reply:
I'm with Veselin on this.

Reply:
Please see my previous comments. I got rid of the ParMultigridBilinearForm and moved it to the examples.